### PR TITLE
Ruby rebuild + cmake => 3.29.2

### PR DESCRIPTION
**Pull request by satmandu**
- Rebuilds `ruby` from `ruby_3_3` branch to fix https://github.com/advisories/GHSA-592j-995h-p23j
- Add `crewlog(string)` function to `crew`.
- Refactor json package removal for upgraded packages.
- Add `save_json` function to safely save the json file.
- Automatically rerun `crew upgrade` when `ruby` is being upgraded.
- Added `CREW_ESSENTIAL_PACKAGES = %w[gcc_lib glibc gmp ruby zlibpkg zstd]` to `const.rb`
- If `distcc` is installed, use `distcc -j` for `CREW_NPROC` calculation.
- grep usage in `commands/remove.rb` needs quoting for file names with spaces.

Works properly:
- [x] `x86_64`
- [x] `i686`
- [x] `armv7l` 

### Run the following to get this pull request's changes locally for testing.
```
CREW_REPO=https://github.com/satmandu/chromebrew.git CREW_BRANCH=ruby_cmake crew update ; yes | crew upgrade -v
```

### CODE REVIEW

**Comment by satmandu** [comment]
Currently there's a problem where `CREW_REPO=https://github.com/satmandu/chromebrew.git CREW_BRANCH=ruby_cmake crew update ; yes | crew upgrade` doesn't work:

```bash
chronos@rpi52-armv7l /usr/local/lib/crew/packages $ CREW_REPO=https://github.com/satmandu/chromebrew.git CREW_BRANCH=ruby_cmake crew update ; yes | crew upgrade
remote: Enumerating objects: 43, done.
remote: Counting objects: 100% (43/43), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 26 (delta 22), reused 26 (delta 22), pack-reused 0
Unpacking objects: 100% (26/26), 15.37 KiB | 215.00 KiB/s, done.
From https://github.com/satmandu/chromebrew
 * branch                ruby_cmake -> FETCH_HEAD
HEAD is now at 111cbde75 cmake => 3.29.2
Package lists, crew, and library updated.
Checking for package updates...
ruby could be updated from 3.3.0 to 3.3.0-2f65458
cmake could be updated from 3.29.0 to 3.29.2
                                                                                                                                                                                                              2 packages can be updated.
Run `crew upgrade` to update all packages or `crew upgrade <package1> [<package2> ...]` to update specific packages.
Updating packages...
ruby: Ruby is a dynamic, open source programming language with a focus on simplicity and productivity.
Performing pre-flight checks...
Precompiled binary available, downloading...
Cannot find cached archive. 😔 Will download.
Ruby archive downloaded.
Unpacking archive using 'tar', this may take a while...
Removing since upgrade or reinstall...
Refusing to remove essential package ruby.
Performing pre-install...
Performing install...
Using rdfind to convert duplicate files to hard links.
Now scanning ".", found 17612 files.
Now have 17612 files in total.
Removed 15 files due to nonunique device and inode.
Total size is 40856842 bytes or 39 MiB
Removed 1997 files due to unique sizes from list. 15600 files left.
Now eliminating candidates based on first bytes: removed 15481 files from list. 119 files left.
Now eliminating candidates based on last bytes: removed 68 files from list. 51 files left.
Now eliminating candidates based on sha1 checksum: removed 51 files from list. 0 files left.
It seems like you have 0 files that are not unique
Totally, 0 B can be reduced.
Now making hard links.
Making 0 links.
Performing post-install for ruby...
Updating ruby gems. This may take a while...
Ruby installed!
Ruby was upgraded. Please run 'crew upgrade' again to make sure upgrades are complete.
chronos@rpi52-armv7l /usr/local/lib/crew/packages $ crew upgrade
Updating packages...
ruby: Ruby is a dynamic, open source programming language with a focus on simplicity and productivity.
Performing pre-flight checks...
Precompiled binary available, downloading...
Archive found in cache
Unpacking archive using 'tar', this may take a while...
Removing since upgrade or reinstall...
Refusing to remove essential package ruby.
Performing pre-install...
Performing install...
Using rdfind to convert duplicate files to hard links.
Now scanning ".", found 17612 files.
Now have 17612 files in total.
Removed 15 files due to nonunique device and inode.
Total size is 40856842 bytes or 39 MiB
Removed 1997 files due to unique sizes from list. 15600 files left.
Now eliminating candidates based on first bytes: removed 15481 files from list. 119 files left.
Now eliminating candidates based on last bytes: removed 68 files from list. 51 files left.
Now eliminating candidates based on sha1 checksum: removed 51 files from list. 0 files left.
It seems like you have 0 files that are not unique
Totally, 0 B can be reduced.
Now making hard links.
Making 0 links.
Performing post-install for ruby...
Updating ruby gems. This may take a while...
Ruby installed!
Ruby was upgraded. Please run 'crew upgrade' again to make sure upgrades are complete.
chronos@rpi52-armv7l /usr/local/lib/crew/packages $
```

File: packages/ruby.rb
Comment (uberhacker) [inline packages/ruby.rb:31]: What is conflicting?  Is it really OK?  If there is another ruby package that has conflicting files, removing that package could brick crew?

File: packages/cmake.rb
Comment (Zopolis4) [inline packages/cmake.rb:23]: Does this really need to be the case?

File: packages/ruby.rb
Comment (satmandu) [reply packages/ruby.rb:31]: Only conflicts with Ruby ...

Comment (satmandu) [reply packages/ruby.rb:31]: Conflict code is wonky ... 🤦‍♂️😔

File: packages/cmake.rb
Comment (satmandu) [reply packages/cmake.rb:23]: Just trying to speed up the build. Tests take forever... Using distcc helps...

**Comment by satmandu** [comment]
Added some debug code which shows this happening:
```bash
Removing since upgrade or reinstall...
Refusing to remove essential package ruby.
Performing pre-install...
Performing install...
Using rdfind to convert duplicate files to hard links.
Now scanning ".", found 17612 files.
Now have 17612 files in total.
Removed 15 files due to nonunique device and inode.
Total size is 43133583 bytes or 41 MiB
Removed 2002 files due to unique sizes from list. 15595 files left.
Now eliminating candidates based on first bytes: removed 15478 files from list. 117 files left.
Now eliminating candidates based on last bytes: removed 68 files from list. 49 files left.
Now eliminating candidates based on sha1 checksum: removed 49 files from list. 0 files left.
It seems like you have 0 files that are not unique
Totally, 0 B can be reduced.
Now making hard links.
Making 0 links.
before json change
{
  "name": "ruby",
  "version": "3.3.0",
  "binary_sha256": "a97a17336b4d233129fb5a55c82d32b38b981d017cda70ba3cbb51012bc129ca"
}
{
  "name": "ruby",
  "version": "3.3.0-2f65458",
  "sha256": "c554a429edbc1d104be2f7a149f91b3892e81605dc550f26a918c2f8bb749c22"
}
after json change
{
  "name": "ruby",
  "version": "3.3.0",
  "binary_sha256": "a97a17336b4d233129fb5a55c82d32b38b981d017cda70ba3cbb51012bc129ca"
}
{
  "name": "ruby",
  "version": "3.3.0-2f65458",
  "sha256": "c554a429edbc1d104be2f7a149f91b3892e81605dc550f26a918c2f8bb749c22"
}
{
  "name": "ruby",
  "version": "3.3.0-2f65458",
  "sha256": "c554a429edbc1d104be2f7a149f91b3892e81605dc550f26a918c2f8bb749c22"
}
Performing post-install for ruby...
```

**Comment by Zopolis4** [comment]
Is this happening for any other package?

File: bin/crew
Comment (uberhacker) [inline bin/crew:1288]: Why is this necessary?  Doesn't it just get added back in again?  Wouldn't a better approach be to check if the package exists and if not, add it?

Comment (satmandu) [reply bin/crew:1288]: A second one gets added in if it isn't first removed.

We need
`If the name is in the array
  remove it
then add name
...
`

Comment (uberhacker) [reply bin/crew:1288]: A second one is added only because you're not checking if it exists beforehand.

Comment (satmandu) [reply bin/crew]: Adjusted code to remove the entry if it is in `CREW_ESSENTIAL_PACKAGES`.

**Comment by satmandu** [comment]
Everything seems to work now.

File: bin/crew
Comment (uberhacker) [inline bin/crew:126]: I sure would like to keep this simple.  Why load twice?

File: bin/crew
Comment (uberhacker) [inline bin/crew:1291]: Still don't quite understand why the package needs to be removed from device.json.

File: bin/crew
Comment (satmandu) [reply bin/crew:126]: Because there is different behavior whether `symbolize_names: true` is set or not. That's the prior behavior, and @Zopolis4 has been doing stuff that works without it being set, so I kept both until we figure out if we actually do need `symbolize_names: true` set everywhere...

File: bin/crew
Comment (uberhacker) [inline bin/crew:1312]: Check if package is already in installed_packages prior to push?

File: lib/const.rb
Comment (uberhacker) [inline lib/const.rb:125]: What's the difference?

File: bin/crew
Comment (satmandu) [reply bin/crew:1291]: Upgrade is the union of remove and install.

Remove takes out the package's line from the json file, except when one of the `CREW_ESSENTIAL_PACKAGES` is being dealt with, in which case remove returns early and the json file never gets updated.

This works around that, as otherwise, a new line gets added to the json file
This is a problem because the version detection for packages looks for the first package info returned from the json file, and since the old info is still there after an upgrade due to this bug, crew thinks that the package still needs to be upgraded after the new package is installed.

File: packages/cmake.rb
Comment (uberhacker) [inline packages/cmake.rb:47]: What happens if cmake is not installed?

File: lib/const.rb
Comment (satmandu) [reply lib/const.rb:125]: With multiple distcc hosts, the number can be much higher than what nproc gives, as work can be offloaded to other machines...

File: bin/crew
Comment (satmandu) [reply bin/crew:1312]: If the package is already installed we still need to delete the package info... We could do a check for that, if that would make more sense?

File: packages/ruby.rb
Comment (uberhacker) [inline packages/ruby.rb:6]: Does it make sense to wait until the next stable release?  I only say this because the maintainers probably have a good reason for not tagging a new release.

**Review by uberhacker — COMMENTED** [review]
See comments.

File: bin/crew
Comment (Zopolis4) [reply bin/crew:126]: We don't need it set at all-- it's just a question of the way we access values (i.e. :name vs 'name')

File: bin/crew
Comment (Zopolis4) [reply bin/crew:1291]: Removing the package should remove the package from the skin file, though.

File: packages/ruby.rb
Comment (Zopolis4) [reply packages/ruby.rb:6]: This is for a security fix, so it might make sense to get this through before a new stable.

File: bin/crew
Comment (Zopolis4) [reply bin/crew:1291]: Ah, I see-- the removal is getting skipped, so the package isn't getting removed.

**Comment by Zopolis4** [comment]
How does upgrading an essential package work then, given that we don't (to my knowledge) remove the currently running ruby binary?

File: bin/crew
Comment (uberhacker) [reply bin/crew:126]: What is the default for symbolize_names?  If we don't need to set it, let's not set it.  I'd be more inclined to want it to be flexible for symbols or names if that is possible.

File: packages/ruby.rb
Comment (satmandu) [reply packages/ruby.rb:6]: Also, this is from their stable branch... the update will trigger a gem rebuild for any vulnerable versions installed.

**Comment by satmandu** [comment]
> How does upgrading an essential package work then, given that we don't (to my knowledge) remove the currently running ruby binary?

Crew should overwrite the package without removing it, thus never hopefully creating a moment when there aren't working binaries for the package on disk. (Which isn't entirely true, especially if doing a massive number of updates and dependencies get mucked up during the update... but that's arguably an edge case.)

File: bin/crew
Comment (satmandu) [reply bin/crew:1311]: I've added a package installed check right before it is added.  (Right now we DO have a situation where the package keeps getting added...)

File: bin/crew
Comment (Zopolis4) [reply bin/crew:126]: It is not set by default (neither is our custom method to symbolize values), and no method for supporting both is available. 

Feel free to change it and all the code that relies on it, or I'll eventually refactor all of the code.

Comment (satmandu) [reply bin/crew:126]: Looks like just flipping it to not set it breaks things:
```
chronos@cheekon-x86_64 /usr/local/lib/crew/packages $ crew upgrade
/usr/local/bin/crew:394:in `upgrade': undefined method `each' for nil (NoMethodError)

    @device[:installed_packages].each do |pkg|
                                ^^^^^
        from /usr/local/bin/crew:1878:in `upgrade_command'
        from /usr/local/bin/crew:1908:in `<main>'
chronos@cheekon-x86_64 /usr/local/lib/crew/packages $
```

Comment (satmandu) [reply bin/crew:126]: uh... well, maybe I can fix that...

Comment (Zopolis4) [reply bin/crew:126]: by and large you'll have to replace :foo with 'foo' whenever json is concerned.

Comment (satmandu) [reply bin/crew:126]: not easily... I think this should be left for a cleanup in a future PR...

Comment (satmandu) [reply bin/crew:126]: @Zopolis4 Going in reverse, code like `@device[:installed_packages].delete_if { |entry| entry['name'] == @pkg.name }` doesn't work. Any idea how to fix that? (That's only used twice, so it might be easier to take care of those...)

Comment (satmandu) [reply bin/crew:126]: I think I got it working...

Comment (satmandu) [reply bin/crew]: Ok I removed it...

File: bin/crew
Comment (uberhacker) [inline bin/crew:446]: Can we move line 441 to this conditional so it makes more logical sense?
```
else
  puts 'Packages have been updated.'.lightgreen
end
```

Comment (satmandu) [reply bin/crew:445]: Done.

**Review by uberhacker — APPROVED** [review]

**Pull request merged**
